README.md
[![Build Status](https://travis-ci.org/ropensci/geonames.png?branch=master)](https://travis-ci.org/ropensci/geonames)

geonames
========

R package for accessing the geonames.org API

Install from here using `devtools`:

 * `install.packages("devtools")` -- if you don't have devtools yet
 * `require(devtools)`
 * `install.packages("rjson")` -- gets this dependency from CRAN
 * `install_github("ropensci/geonames")`

A version of this will be pushed to CRAN.

To enable all of the functions of this package, register for a geonames.org username at http://www.geonames.org/login/ and then enable access to the geonames free webservices by clicking http://www.geonames.org/enablefreewebservice Once that's done, in your `R` session, run `options(geonamesUsername="myusername")` to authenticate with the webservice. More information about the parameters for many of the functions in this package can be found here: http://www.geonames.org/export/ws-overview.html

[![Part of ROpenSci](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)


NEWS
geonames 0.999 (2019-02-17)
=============================

### CHANGES

 * Geonames changed their service endpoint host name (#16)

 * Set hostname on load instead of attach (#15)



inst/tests/testing.R
###
### some sample usages
###

### note these aren't in the automated test directory because
### they rely on internet connectivity and a good connection to
### the geonames server.

GNchildren(3175395)

GNcities(north=44.1,south=-9.9,east=-22.4,west=55.2,lang="de")

GNcountryCode(lat=47.03,lng=10.2)

GNcountryInfo()
GNcountryInfo("DE")


GNearthquakes(north=44.1,south=-9.9,east=-22.4,west=55.2)

GNfindNearByWeather(57,-2)

GNfindNearbyStreets(37.45,-122.18)

GNwikipediaSearch("london")
GNfindNearbyWikipedia(postalcode=8775,country="CH",radius=10)
GNwikipediaBoundingBox(north=44.1,south=-9.9,east=-22.4,west=55.2)

GNtimezone(57.01,-2)
GNtimezone(lat=0,lng=-40)

# new radius functionality
GNtimezone(lat=0,lng=-40, radius=200)

GNfindNearbyPostalCodes(lat=47,lng=9)
GNpostalCodeSearch(postalcode=90210,country="FI")
GNpostalCodeSearch(postalcode=90210,country="US")
GNpostalCodeLookup(postalcode="LA1",country="UK")
GNpostalCodeLookup(postalcode="90210")

GNsearch(q="london",maxRows=10)

GNneighbours(3041565)

GNneighbourhood(40.7834,-73.96625)
GNpostalCodeCountryInfo()

# this caused warnings with the timezone being a list:
GNfindNearbyPlaceName(52,-128,300, "30","FULL") 


R/postal.R

##
## postal codes...
##

##' find postal code by lat long or code
##'
##' find postal code
##'
##' API doc for GNfindNearbyPostalCodes is at \url{http://www.geonames.org/export/web-services.html#findNearbyPostalCodes}
##' 
##' @title find postal code
##' @param ... search parameters, see geonames web docs for details
##' @return postal code records
##' @export
##' @author Barry Rowlingson
GNfindNearbyPostalCodes=function(...){
  return(gnRaggedDataFrame("findNearbyPostalCodesJSON",list(...),"postalCodes"))
}


##' postal code lookup
##'
##' postal code lookup
##'
##' API doc for GNpostalCodeLookup is at \url{http://www.geonames.org/export/web-services.html#postalCodeLookupJSON}
##' 
##' @title postal code lookup
##' @param ... parameters
##' @return list of places for a given input postal code
##' @export
##' @author Barry Rowlingson
GNpostalCodeLookup=function(...){
  # only defined for some countries...
  return(gnRaggedDataFrame("postalCodeLookupJSON",list(...),"postalcodes"))
}

##' search for postal code
##'
##' full text search for postal codes
##'
##' API doc for GNpostalCodeSearch is at \url{http://www.geonames.org/export/web-services.html#postalCodeSearch}
##' 
##' @title postal code search
##' @param ... search parameters
##' @return postal code record
##' @export
##' @author Barry Rowlingson
GNpostalCodeSearch=function(...){
  return(gnRaggedDataFrame("postalCodeSearchJSON",list(...),"postalCodes"))
}

##' countries with postal code info
##'
##' list countries with postal code info
##'
##' API doc for GNpostalCodeCountryInfo is at \url{http://www.geonames.org/export/web-services.html#postalCodeCountryInfo}
##' 
##' @title postal code info
##' @return list of countries with postal codes on geonames
##' @author Barry Rowlingson
##' @export
GNpostalCodeCountryInfo=function(){
#
### now uses the JSON version - used to parse XML here!
###
  return(gnRaggedDataFrame("postalCodeCountryInfoJSON",list(),"geonames"))
}


R/search.R

##
## searching...
##

##' search geonames
##'
##' general search call
##' 
##' API doc for GNsearch is at \url{http://www.geonames.org/export/geonames-search.html}
##' 
##' @title search geonames
##' @param ... search parameters
##' @return matched records
##' @examples
##' \dontrun{
##' # Find places called Lancaster and 'geocode' the result
##' (lanc_df <- GNsearch(name = "Lancaster", country = "UK"))
##' lanc_coords <- lanc_df[1, c("lng", "lat")]
##' }
##' @export
##' @author Barry Rowlingson
GNsearch=function(...){
  return(gnRaggedDataFrame("searchJSON",list(...),"geonames"))
}

##' find neighbourhood
##'
##' find neighbourhood
##' 
##' API doc for GNneighbourhood is at \url{http://www.geonames.org/export/web-services.html#neighbourhood}
##' 
##' @title neighbourhood
##' @param lat latitude
##' @param lng longitude
##' @return neighbourhood records
##' @export
##' @author Barry Rowlingson
GNneighbourhood=function(lat,lng){
# US cities only
  return(getJson("neighbourhoodJSON",list(lat=lat,lng=lng))$neighbourhood)
}

##' find nearby entities
##'
##' nearby search 
##' 
##' API doc for GNfindNearby is at \url{http://www.geonames.org/export/web-services.html#findNearby}
##' 
##' @title nearby search
##' @param ... search parameters
##' @return matched records
##' @export
##' @author Barry Rowlingson
GNfindNearby=function(...){
  warning("Not documented properly yet by geonames")
  return(getJson("findNearbyJSON",list(...)))
}

##' find nearby populated place
##'
##' search for populated places
##' 
##' API doc for GNfindNearbyPlaceName is at \url{http://www.geonames.org/export/web-services.html#findNearbyPlaceName}
##' 
##' @title populated place search
##' @param lat latitude
##' @param lng longitude
##' @param radius search radius
##' @param maxRows max records returned
##' @param style verbosity of record
##' @return nearby populated place records
##' @author Barry Rowlingson
##' @export
GNfindNearbyPlaceName=function(lat,lng,radius="",maxRows="10",style="MEDIUM"){
  return(gnDataFrame("findNearbyPlaceNameJSON",list(lat=lat,lng=lng,radius=radius,style=style,maxRows=maxRows),"geonames"))
}

##' find nearby streets (US only)
##'
##' for a lat-long, find nearby US streets
##' 
##' API doc for GNfindNearbyStreets is at \url{http://www.geonames.org/maps/us-reverse-geocoder.html#findNearbyStreets}
##' 
##' @title nearby street finding
##' @param lat latitude
##' @param lng longitude
##' @return street records
##' @export
##' @author Barry Rowlingson
GNfindNearbyStreets=function(lat,lng){
  return(gnDataFrame("findNearbyStreetsJSON",list(lat=lat,lng=lng),"streetSegment"))
}

##' find nearest street and address
##'
##' search US for nearest street and address
##' 
##' API doc for GNfindNearestAddress is at \url{http://www.geonames.org/maps/us-reverse-geocoder.html#findNearestAddress}
##' 
##' @title nearest address
##' @param lat latitude
##' @param lng longitude
##' @return address record
##' @export
##' @author Barry Rowlingson
GNfindNearestAddress=function(lat,lng){
  return(as.data.frame(getJson("findNearestAddressJSON",list(lat=lat,lng=lng))$address))
}

##' search US for nearest intersection
##'
##' finds nearest intersection
##'
##' API doc for GNfindNearestIntersection is at \url{http://www.geonames.org/maps/us-reverse-geocoder.html#findNearestIntersection}
##' 
##' @title nearest intersection 
##' @param lat latitude
##' @param lng longitude
##' @return intersection record
##' @export
##' @author Barry Rowlingson
GNfindNearestIntersection=function(lat,lng){
  return(as.data.frame(getJson("findNearestIntersectionJSON",list(lat=lat,lng=lng))$intersection))
}



R/general.R

##
## General Info
##
##' Get country code
##'
##' country code for location
##' 
##' API doc for GNcountryCode is at \url{http://www.geonames.org/export/web-services.html#countrycode}
##' 
##' @title country code for location
##' @param lat latitude
##' @param lng longitude
##' @param lang language code
##' @param radius radius size
##' @return country record
##' @export
##' @author Barry Rowlingson
GNcountryCode=function(lat,lng,lang="",radius=""){
# returns a single country
    return(getJson("countryCode",list(lat=lat,lng=lng,radius=radius,lang=lang,type="JSON")))
}

##' Get country info
##'
##' country info
##'
##' API doc for GNcountryInfo is at \url{http://www.geonames.org/export/web-services.html#countryInfo}
##' 
##' @title country info
##' @param country country code
##' @param lang language code
##' @return country record info
##' @author Barry Rowlingson
##' @export
GNcountryInfo=function(country="",lang=""){
# can return many records if no country code given
  return(gnDataFrame("countryInfo",list(country=country,lang=lang,type="JSON"),"geonames"))
}

##' Search for cities
##'
##' find cities
##' 
##' API doc for GNcities is at \url{http://www.geonames.org/export/JSON-webservices.html#citiesJSON}
##' 
##' @title find cities
##' @param north north bound
##' @param east east bound
##' @param south south bound
##' @param west west bound
##' @param lang language code
##' @param maxRows max number of records to return
##' @return city records
##' @export
##' @author Barry Rowlingson
GNcities=function(north,east,south,west,lang="en",maxRows=10){
  return(gnDataFrame("citiesJSON",list(north=north,east=east,west=west,south=south,lang=lang,maxRows=maxRows),"geonames"))
}

##' get timezone
##'
##' timezone for location
##' 
##' API doc for GNtimezone is at \url{http://www.geonames.org/export/web-services.html#timezone}
##' 
##' @title timezone for location
##' @param lat latitude
##' @param lng longitude
##' @param radius search radius
##' @return time zone record
##' @author Barry Rowlingson
##' @export
GNtimezone=function(lat,lng, radius=0){
  return(as.data.frame(getJson("timezoneJSON",list(lat=lat,lng=lng, radius=0))))
}

##' country code and admin subdivision
##'
##' looks up country and admin subdivisions
##' 
##' API doc for GNcountrySubdivision is at \url{http://www.geonames.org/export/web-services.html#countrysubdiv}
##' 
##' @title country code and subdivision
##' @param lat latitude
##' @param lng longitude
##' @param lang language code
##' @param radius search radius
##' @param maxRows max number of returned records
##' @return iso country code
##' @export
##' @author Barry Rowlingson
GNcountrySubdivision=function(lat,lng,lang="en",radius="",maxRows=10){
# could return >1 record, but only the XML version seems to at the moment...
  return(getJson("countrySubdivisionJSON",list(lat=lat,lng=lng,lang=lang,radius=radius,maxRows=maxRows)))
}



R/geonames.R

##     Copyright (C) 2008 Barry Rowlingson

##     This program is free software: you can redistribute it and/or modify
##     it under the terms of the GNU General Public License as published by
##     the Free Software Foundation, either version 3 of the License, or
##     (at your option) any later version.

##     This program is distributed in the hope that it will be useful,
##     but WITHOUT ANY WARRANTY; without even the implied warranty of
##     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##     GNU General Public License for more details.

##     You should have received a copy of the GNU General Public License
##     along with this program.  If not, see <http://www.gnu.org/licenses/>.

#' Query the geonames web API for geographic data
#'
#' www.geonames.org is a service where you can query for global geographic
#' data such as administrative areas, populated places, weather data etc.
#'
#' The functions in this package are mostly thin wrappers to the API
#' calls documented at the geonames web services
#' overview \url{http://www.geonames.org/export/ws-overview.html}.
#'
#' A set of example calls are supplied in a file with the package. Once
#' you have set your geonames username with
#' \code{options(geonamesUsername="myusernamehere")} you can run
#' this with \code{source(system.file("tests","testing.R",package="geonames"),echo=TRUE)}
#' 
#' @docType package
#' @name geonames
NULL

.onLoad = function(libname,pkgname){
  ## sometimes they take this down and change it...
  if(is.null(getOption("geonamesHost"))){
    options(geonamesHost="api.geonames.org")
  }
}

##
## useful functions
##

getJson=function(name,plist){
#
# call a geonames JSON service with named args from plist
#
  url=paste("http://",options()$geonamesHost,"/",name,"?",sep="")
  if(!is.null(options()$geonamesUsername)){
    plist[["username"]]=options()$geonamesUsername
  }else{
      warning("No geonamesUsername set. See http://geonames.wordpress.com/2010/03/16/ddos-part-ii/ and set one with options(geonamesUsername=\"foo\") for some services to work")
  }
    
  olist = list()
  for(p in 1:length(plist)){
    olist[[p]]=paste(names(plist)[p],"=",utils::URLencode(as.character(plist[[p]]),reserved=TRUE),sep="")
  }
  pstring=paste(unlist(olist),collapse="&")
  url=paste(url,pstring,sep='')  
  u=url(url,open="r")
  d=readLines(u,warn=FALSE)
  close(u)
  data = rjson::fromJSON(d)
  if(length(data$status)>0){
    stop(paste("error code ",data$status$value," from server: ",data$status$message,sep=""))
  }
  return(data)
}


gnDataFrame=function(name,params,ename){
#
# return a data frame constructed from a JSON call
# 

###
### this code tried to be more efficient but each column was a list.
### eventually I'll figure out a better way to construct a data frame, until then
### i'll just use the ragged version.
#  json = getJson(name,params)
#  return(as.data.frame(do.call("rbind",json[[ename]])))

### another poss is:  do.call("rbind",lapply(l,data.frame))
### but it errors if the data frame is ragged. We could test for this and
### use gnRaggedDataFrame if it fails
  
  return(gnRaggedDataFrame(name,params,ename))

}

gnRaggedDataFrame=function(name,params,ename){
#
# if the return is a Json list with items of different structure but we want a data frame,
# use this and it will fill missing items with <NA> values.
#
  json = getJson(name,params)[[ename]]
  names=NULL
  for(j in json){
    names=unique(c(names,names(unlist(j))))
  }

  m=data.frame(matrix(NA,ncol=length(names),nrow=length(json)))
  names(m) = names
  row=1
  for(j in json){
    for(ename in names(unlist(j))){
       m[row,ename]=unlist(j)[[ename]]
    }
    row=row+1
  }
  return(m)
}






gnerrorCodes = function(){
  errorcodes = list()
  errorcodes[["10"]]="Authorization Exception"
  errorcodes[["11"]]="record does not exist"
  errorcodes[["12"]]="other error"
  errorcodes[["13"]]="database timeout"
  errorcodes[["14"]]="invalid parameter"
  errorcodes[["15"]]="no result found"
  errorcodes[["16"]]="duplicate exception"
  errorcodes[["17"]]="postal code not found"
  errorcodes[["18"]]="daily limit of credits exceeded"
  errorcodes[["19"]]="hourly limit of credits exceeded"
  errorcodes[["20"]]="weekly limit of credits exceeded"
  return(errorcodes)
}

gnerrorString=function(code){
  code=as.character(code)
  allCodes = gnerrorCodes()
  if(!is.null(allCodes[[code]])){
    return(allCodes[[code]])
  }else{
    return(paste("Unknown Code: ",code,sep=""))
  }
}



R/hierarchy.R
#' Admin area hierarchy
#'
#' See \url{http://www.geonames.org/export/ws-overview.html}
#' for a full description of valid arguments and return values
#'
#'  API doc for GNchildren is at \url{http://www.geonames.org/export/place-hierarchy.html#children}
#' 
#' API doc for GNhierarchy is at \url{http://www.geonames.org/export/place-hierarchy.html#hierarchy}
#' 
#' API doc for GNsiblings is at \url{http://www.geonames.org/export/place-hierarchy.html#siblings}
#' 
#' API doc for GNneighbours is at \url{http://www.geonames.org/export/place-hierarchy.html#neighbours}
#' 
#' @param geonameId a geonames ID value
#' @param ... other parameters to pass to geonames
#' @name hierarchy

#' @rdname hierarchy
#' @export
GNchildren=function(geonameId,...){
# allows name, lang, others?
  return(gnDataFrame("childrenJSON",list(geonameId=geonameId,...),"geonames"))
}

#' @rdname hierarchy
#' @export
GNhierarchy=function(geonameId,...){
  return(gnRaggedDataFrame("hierarchyJSON",list(geonameId=geonameId,...),"geonames"))
}

#' @rdname hierarchy
#' @export
GNsiblings=function(geonameId,...){
  return(gnDataFrame("siblingsJSON",list(geonameId=geonameId,...),"geonames"))
}

#' @rdname hierarchy
#' @export
GNneighbours=function(geonameId,...){
# works for countries only
  return(gnDataFrame("neighboursJSON",list(geonameId=geonameId,...),"geonames"))
}



R/topography.R

##
## Topography (heights)
##

##' height from topo30
##'
##' get height from topo30 data
##' 
##' API doc for GNgtopo30 is at \url{http://www.geonames.org/export/web-services.html#gtopo30}
##'
##' @title topo30 height
##' @param lat latitude
##' @param lng longitude
##' @return height record
##' @examples
##' \dontrun{
##' GNgtopo30(lat=54,lng=-1)
##' }
##' @export
##' @author Barry Rowlingson
GNgtopo30=function(lat,lng){
  return(as.data.frame(getJson("gtopo30JSON",list(lat=lat,lng=lng))))
}

##' height from srtm3 data
##'
##' get srtm3 height
##' 
##' API doc for GNsrtm3 is at \url{http://www.geonames.org/export/web-services.html#srtm3}
##' 
##' @title srtm3 height
##' @param lat latitude
##' @param lng longitude
##' @return height record
##' @examples
##' \dontrun{
##' GNsrtm3(lat=54,lng=-1)
##' }
##' @export
##' @author Barry Rowlingson
GNsrtm3=function(lat,lng){
  return(as.data.frame(getJson("srtm3JSON",list(lat=lat,lng=lng))))
}



R/environment.R

##
## weather, earthquake, giant radioactive lizard invasion:
##

##' get weather at location
##'
##' get weather
##' 
##' API doc for GNfindNearByWeather is at \url{http://www.geonames.org/export/JSON-webservices.html#findNearByWeatherJSON}
##' 
##' @title weather at location
##' @param lat latitude
##' @param lng longitude
##' @return weather record
##' @examples
##' \dontrun{
##' GNfindNearByWeather(57,-2)
##' }
##' @export
##' @note check capitalisation of 'NearBy'
##' @author Barry Rowlingson
GNfindNearByWeather=function(lat,lng){
  return(getJson("findNearByWeatherJSON",list(lat=lat,lng=lng))$weather)
}

##' weather stations in region
##'
##' get weather stations in region with latest readings
##' @title weather stations in box
##' @param north north bound
##' @param east east bound
##' @param south south bound
##' @param west west bound
##' @param maxRows max records to return
##' @return weather records
##' @export
##' @author Barry Rowlingson
GNweather=function(north,east,south,west,maxRows=10){
  return(gnRaggedDataFrame("weatherJSON",list(north=north,east=east,west=west,south=south,maxRows=maxRows),"weatherObservations"))
}

##' weather record from ICAO station
##'
##' get most recent ICAO station data
##' 
##' API doc for GNweatherIcao is at \url{http://www.geonames.org/export/JSON-webservices.html#weatherIcaoJSON}
##' 
##' @title ICAO weather station data
##' @param ICAO ICAO code
##' @return weather record
##' @export
##' @author Barry Rowlingson
GNweatherIcao=function(ICAO){
  return(as.data.frame(getJson("weatherIcaoJSON",list(ICAO=ICAO))$weatherObservation))
}

##' recent earthquakes
##'
##' get recent earthquakes in a region
##'
##' API doc for GNearthquakes is at \url{http://www.geonames.org/export/JSON-webservices.html#earthquakesJSON}
##' 
##' @title recent earthquakes
##' @param north  north bound
##' @param east east bound
##' @param south south bound
##' @param west west bound
##' @param date optional date
##' @param minMagnitude optional minimal magnitude
##' @param maxRows max records to return
##' @return earthquake records
##' @examples
##' \dontrun{
##' GNearthquakes(north=44.1,south=-9.9,east=-22.4,west=55.2)
##' }
##' @export
##' @author Barry Rowlingson
GNearthquakes=function(north,east,south,west,date,minMagnitude,maxRows=10){
  params = list(north=north,south=south,east=east,west=west,maxRows=maxRows)
  if(!missing(date)){ 
    params$date=date
  }
  if(!missing(minMagnitude)){
    params$minMagnitude=minMagnitude
  }
  return(gnDataFrame("earthquakesJSON",params,"earthquakes"))
}



R/wikipedia.R

##
## wikipedia searching
##

##' find nearby wikipedia entries
##'
##' search wikipedia entries by lat/lng or location name parameters
##' 
##' API doc for GNfindNearbyWikipedia is at \url{http://www.geonames.org/export/wikipedia-webservice.html#findNearbyWikipedia}
##' @title nearby wikipedia entries
##' @param ... see geonames.org documentation
##' @return wikipedia entries
##' @examples
##' \dontrun{
##' GNfindNearbyWikipedia(postalcode=8775,country="CH",radius=10)
##' }
##' @export
##' @author Barry Rowlingson
GNfindNearbyWikipedia=function(...){
  return(gnRaggedDataFrame("findNearbyWikipediaJSON",list(...),"geonames"))
}

##' wikipedia articles in bounding box
##'
##' find articles in a box
##' 
##' API doc for GNwikipediaBoundingBox is at \url{http://www.geonames.org/export/wikipedia-webservice.html#wikipediaBoundingBox}
##' 
##' @title wikipedia articles in a box
##' @param ... parameters (north, south, east, west etc.)
##' @return wikipedia records
##' @examples
##' \dontrun{
##' GNwikipediaBoundingBox(north=44.1,south=-9.9,east=-22.4,west=55.2)
##' }
##' @export
##' @author Barry Rowlingson
GNwikipediaBoundingBox=function(...){
  return(gnRaggedDataFrame("wikipediaBoundingBoxJSON",list(...),"geonames"))
}


##' wikipedia fulltext search
##'
##' find geolocated articles in wikipedia
##' 
##' API doc for GNwikipediaSearch is at \url{http://www.geonames.org/export/wikipedia-webservice.html#wikipediaSearch}
##' 
##' @title search wikipedia
##' @param q search string
##' @param maxRows maximum returned records
##' @return wikipedia entries
##' @examples
##' \dontrun{
##' GNwikipediaSearch("london")
##' }
##' @export
##' @author Barry Rowlingson
GNwikipediaSearch=function(q,maxRows=10){
  return(gnRaggedDataFrame("wikipediaSearchJSON",list(q=q,maxRows=maxRows),"geonames"))
}
